domain/cloudimagemetadata/service/service.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/collections/set"
+
+	"github.com/juju/juju/core/cloudimagemetadata"
+	"github.com/juju/juju/domain/cloudimagemetadata/errors"
+	jujuerrors "github.com/juju/juju/internal/errors"
+)
+
+// State is an interface of the persistence layer for managing cloud image metadata
+type State interface {
+	// SaveMetadata saves a list of cloud image metadata to the state.
+	SaveMetadata(ctx context.Context, metadata []cloudimagemetadata.Metadata) error
+
+	// DeleteMetadataWithImageID removes the metadata associated with the given imageID from the state.
+	DeleteMetadataWithImageID(ctx context.Context, imageID string) error
+
+	// FindMetadata retrieves a list of cloud image metadata based on the provided criteria.
+	// It returns the matched metadata or a [github.com/juju/juju/domain/cloudimagemetadata/errors.NotFound] error
+	// If there are no one.
+	FindMetadata(ctx context.Context, criteria cloudimagemetadata.MetadataFilter) ([]cloudimagemetadata.Metadata, error)
+
+	// AllCloudImageMetadata retrieves all cloud image metadata from the state and returns them as a list.
+	AllCloudImageMetadata(ctx context.Context) ([]cloudimagemetadata.Metadata, error)
+
+	// SupportedArchitectures returns a set of strings representing the architectures supported by the cloud images.
+	SupportedArchitectures(ctx context.Context) set.Strings
+}
+
+// Service provides the API for working with cloud image metadata
+type Service struct {
+	st State
+}
+
+// NewService returns a new service around the persistence layer
+func NewService(st State) *Service {
+	return &Service{
+		st: st,
+	}
+}
+
+// SaveMetadata saves the provided cloud image metadata if non-empty and valid.
+// It returns a [errors.NothingToSave] if the input are empty, or a [errors.Invalid] if at least one of the inputs are
+// invalid.
+func (s Service) SaveMetadata(ctx context.Context, metadata []cloudimagemetadata.Metadata) error {
+	if len(metadata) == 0 {
+		return errors.NothingToSave
+	}
+
+	if err := s.validateAllMetadata(metadata); err != nil {
+		return err
+	}
+
+	return s.st.SaveMetadata(ctx, metadata)
+}
+
+// DeleteMetadataWithImageID removes all the metadata associated with the given imageID from the state.
+func (s Service) DeleteMetadataWithImageID(ctx context.Context, imageId string) error {
```

Comment:
Question: does the imageId conform to any layout? If it does we could validate it. If not, it doesn't matter.

Reply:
As far I remember, I have seen numeric id. But since in legacy we don't check it, wrongly infers the layout may cause random breaking changes, unless this lay under some convention. Will check on simplestream whenever i will figure out how to. If you have any hints, you would be welcome.

Reply:
Probably as the minimum, just check it's not empty.

Reply:
Deploy ubuntu on AWS && lxd, and i get as ImageId: `ami-<some random characters>` (ami stands for Amazon  Machine Image IMHO)

I don't know if all imageID would follows this pattern, in fact I would be surprised.

ImageID is checked against emptyness in save. I will add an error case here too.

Reply:
That'll do for now then.